### DISCUSSION TITLE
Response not from requested slave (using RTU)

### DISCUSSION THREAD

**Discussion by BigBennnni** [eModbus/eModbus#307] [Q&A]
Hey there,
I'm currently working on a projekt, where I neet to setup an ESP32 as a modbus server/slave on an RTU network. As a master for my tests I'm using a computer with the programm QModMaster as the modbus master. Connected is the ESP32 through a microUSB cable directly to the computer.
The problem that I currently experience is that some of the send packets resume in the error "Response not from requested slave". This happens mostly using the Function Code Write Multiple Registers (1600 packets send 42 errors received). It rarely happens during Read Holding Registers (2000 packets send 1 or 2 errors received). It also doesnt happen in regular intervals like after 20 packets send, it can happen in under 10 packets or I am able to send 30 packets without an error.
As an attempt to debug it I tried a new ESP32, changed the cable, used a scope to check the connection and shortend my code to a bare minimum to be sure my other code isn't the problem. Changing the baudrate also doesn't solve the problem. As this is the only slave on the bus there shouldnt be an issue with collision (I tried to sent a packet every second but still the same)

The code I'm using you can see here:
```
#include <Arduino.h>
#include <SoftwareSerial.h>

// Modbus server include
#include "ModbusServerRTU.h"
#include "ModbusMessage.h"

EspSoftwareSerial::UART debugSerial;

// Create a ModbusRTU server instance listening on Serial with 2000ms timeout
ModbusServerRTU MBserver(Serial, 2000);

const int anzahl_register = 64;
uint16_t memo[anzahl_register];                     // Test server memory: 64 words

ModbusMessage empty;                   // Empty message for initializers

bool anfrage_im_bereich (int addr, int wrds) {
  if ((addr + wrds) <= anzahl_register) {
    return true;
  }
  else {
    return false;
  }
}

// Worker function for function code 0x03
ModbusMessage FC03(ModbusMessage request) {
  uint16_t addr = 0;        // Start address to read            Start Address         wenn 1 dann 1
  uint16_t wrds = 0;        // Number of words to read          Number of Registers   wenn 1 dann 1
  ModbusMessage response;

  // Get addr and words from data array. Values are MSB-first, get() will convert to binary
  request.get(2, addr, wrds);

  if (anfrage_im_bereich(addr, wrds) != true) {
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
    return response;
  }

  response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(wrds * 2));

  // Loop over all words to be sent
  for (uint16_t i = 0; i < wrds; i++) {
    response.add((uint16_t)((request.getServerID() == 1) ? memo[addr + i] : ~memo[addr + i]));
  }
  // Return the data response
  return response;
}


ModbusMessage FC10(ModbusMessage request) {
  uint16_t addr = 0;
  uint16_t wrds = 0;
  ModbusMessage response;

  request.get(2, addr);
  request.get(4, wrds);

  for (uint16_t i = 0; i < wrds; i++) {
    response.add((uint16_t)((request.getServerID() == 1) ? memo[addr + i] : ~memo[addr + i]));
  }

  return ECHO_RESPONSE;
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  //debugSerial.begin(9600, SWSERIAL_8N1, 13, 12, false);
  Serial.begin(115200);
  while (!Serial) {
    ;
  }
  Serial.println("__ OK __");
  
// Register served function code worker for server 1
  MBserver.registerWorker(0x01, READ_HOLD_REGISTER, &FC03);
  MBserver.registerWorker(0x01, WRITE_MULT_REGISTERS, &FC10);


// Start ModbusRTU background task
  MBserver.start();
}

void loop() {
  
}
```

Help with this problem would be very appreciated, because I'm running out of ideas ^^
Thanks in advance 
BigBennnni

**Comment by Miq1** [comment]
Deutsch, vermute ich? 😁 

Using ``Serial`` for both monitor printouts and the RTU server is slightly dangerous. You never know how the core may interfere with the Modbus communication. I also take from your code that you do not use the most recent eModbus library (1.7). If you happen to have upgraded the core already to some 2.x version, I strongly advise to upgrade eModbus as well and use the new UART handling as is described in the docs. Even better with UART1 or UART2 and a USB-to-UART adapter to get out of the way of the monitor handling.
